Code/src/main/java/nl/utwente/group10/ui/components/blocks/function/FunctionBlock.java

```diff
@@ -0,0 +1,248 @@
+        
+        // Create an anchor for the result
+        output = new OutputAnchor(this);
+        outputSpace.setCenter(output);
+        
+        // Make sure the prefWidth is correctly updated.
+        this.prefWidthProperty().bind(functionInfo.widthProperty().add(argumentSpace.prefWidthProperty()));
+        
+        
+        this.setMinSize(Region.USE_PREF_SIZE, Region.USE_PREF_SIZE);
+        this.setMaxSize(Region.USE_PREF_SIZE, Region.USE_PREF_SIZE);
+        functionInfo.setMinWidth(Region.USE_PREF_SIZE);
+        functionInfo.setMaxWidth(Region.USE_PREF_SIZE);
+        
+        // Since at this point of the width of the Labels is unknown, we have to ask for another layout pass.
+        Platform.runLater(this::updateLayout);
+    }
+    
+    /** Updates the layout, if this Pane has a parent. */
+    public void updateLayout() {
+        if (this.getParent() != null) {
+            this.getParent().requestLayout();
+        }
+    }
+    
+    /** @return the name of this FunctionBlock. */
+    public final String getName() {
+        return name.get();
+    }
+
+    /** Sets the name of this FunctionBlock. */
+    public void setName(String name) {
+        this.name.set(name);
+    }
+    
+    /** @return the StringProperty for the name of the function. */
+    public final StringProperty nameProperty() {
+        return name;
+    }
+
+    /** @return the knot index of this FunctionBlock. */
+    public final Integer getKnotIndex() {
+        return argumentSpace.knotIndexProperty().get();
+    }
+
+    /** Sets the new knot index for this FunctionBlock. */
+    public final void setKnotIndex(int index) {
+        argumentSpace.setKnotIndex(index);
+    }
+
+    @Override
+    public final List<InputAnchor> getAllInputs() {
+        return argumentSpace.getInputArguments().stream().map(a -> a.getInputAnchor()).collect(Collectors.toList());
+    }
+
+    /**
+     * @return Only the active (as specified by the knot index) inputs.
+     */
+    @Override
+    public List<InputAnchor> getActiveInputs() {
+        return getAllInputs().subList(0, getKnotIndex());
+    }
+
+    /**
+     * @return InputAnchor with the given index.
+     */
+    public InputAnchor getInput(int index) {
+        return getAllInputs().get(index);
+    }
+
+    @Override
+    public OutputAnchor getOutputAnchor() {
+        return output;
+    }
+
+    /**
+     * On top of updating the expression, this method also adds a record to the
+     * CustomUIPane that maps the expr to this block.
+     */
+    @Override
+    public final void updateExpr() {
+        getPane().removeExprToFunction(expr);
+        expr = new Ident(getName());
+        getPane().putExprToFunction(expr, this);
+        
+        for (InputAnchor in : getActiveInputs()) {
+            expr = new Apply(expr, in.getExpr());
+        }
+        super.updateExpr();
+    }
+    
+    @Override
+    public void invalidateVisualState() {
+        super.invalidateVisualState();
```

Comment:
During testing I used print statements in the super method, so I would suggest leaving the super calls.